### PR TITLE
py-jsonschema and deps: add format_nongpl, fix darwin<15

### PR DESCRIPTION
**Pull request by johankytt**
#### Description

Add the `format_nongpl` variant required for Jupyter 4
- new ports: py-fqdn, py-isoduration, py-uri-template
- add py312: py-jsonpointer, py-rfc3339-validator, py-rfc3986-validator

Attempt to fix the build for OSX Yosemite and older
No old OS available for actual testing of the fix
https://trac.macports.org/ticket/69087

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 14.2.1 23C71 arm64
Xcode 15.2 15C500b

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: python/py-fqdn/Portfile
Comment (reneeotten) [inline python/py-fqdn/Portfile:14]: this is the default for Python 3.7+ subports, so it will be always used here ; please remove

File: python/py-fqdn/Portfile
Comment (reneeotten) [inline python/py-fqdn/Portfile:11]: it the package does not depend on the architecture and it installs the same files on all macOS versions, which typically can be checked from PyPI if the downloadable .whl files are names `*none-any.whl`.
In that case, you should also set `platforms {darwin any}`

File: python/py-isoduration/Portfile
Comment (reneeotten) [inline python/py-isoduration/Portfile:10]: same as above

File: python/py-isoduration/Portfile
Comment (reneeotten) [inline python/py-isoduration/Portfile:13]: same as above

File: python/py-isoduration/Portfile
Comment (reneeotten) [inline python/py-isoduration/Portfile:14]: you will need to specify a `license`

File: python/py-jsonpointer/Portfile
Comment (reneeotten) [inline python/py-jsonpointer/Portfile:30]: these lines should be removed, not commented out. Since these are all Python 3.7+ subports it will use `python.pep517 yes` with the default `setuptools` backend; so, indeed the `py-setuptools` dependency will be added automatically and doesn't need to be specified explicitly here anymore.

File: python/py-jsonschema/Portfile
Comment (reneeotten) [inline python/py-jsonschema/Portfile:12]: why this change?

File: python/py-jsonschema/Portfile
Comment (reneeotten) [inline python/py-jsonschema/Portfile:55]: this should be indented properly with four spaces

File: python/py-jsonschema/Portfile
Comment (reneeotten) [inline python/py-jsonschema/Portfile:56]: this check should be done as usual: ```if {${os.platform} eq "darwin" && ${os.major} < XY {```

File: python/py-jsonschema/Portfile
Comment (reneeotten) [inline python/py-jsonschema/Portfile:70]: the use of `variants` is recommended against unless really necessary. The problem is that you cannot directly depend on a variant, but will need to use the `active_variant` PortGroup which will only check if a port is installed with a variant and, if not, will just throw an error. So if you need to depend on this for another port it should likely not be done with a non-default variant, otherwise the buildbots will never be able to build that port.

File: python/py-uri-template/Portfile
Comment (reneeotten) [inline python/py-uri-template/Portfile:14]: this is the default, please remove

File: python/py-uri-template/Portfile
Comment (reneeotten) [inline python/py-uri-template/Portfile:11]: same as above

File: python/py-uri-template/Portfile
Comment (reneeotten) [inline python/py-uri-template/Portfile:13]: please do not EOL Python versions unless absolutely required

**Review by reneeotten — CHANGES_REQUESTED** [review]

File: python/py-fqdn/Portfile
Comment (johankytt) [reply python/py-fqdn/Portfile:14]: I've seen different behaviour with and without `pep517 yes` present with other ports, but I'll remove it

File: python/py-isoduration/Portfile
Comment (johankytt) [reply python/py-isoduration/Portfile:14]: There's no license listed in the source package and macports guide does not list an option for "missing" or "unknown" license.

File: python/py-jsonschema/Portfile
Comment (johankytt) [reply python/py-jsonschema/Portfile:12]: Because it will install a different version of the port (hence different files) based on the OS version

File: python/py-jsonschema/Portfile
Comment (johankytt) [reply python/py-jsonschema/Portfile:55]: I don't see anything that's not indented to 4 spaces

File: python/py-fqdn/Portfile
Comment (reneeotten) [reply python/py-fqdn/Portfile:14]: Yes, that is for historical reasons and no longer needed after the change in the `python` PortGroup where it is now set as default for Python 3.7+

File: python/py-jsonschema/Portfile
Comment (reneeotten) [reply python/py-jsonschema/Portfile:55]: sorry, I missed that the first `if`-statement actually as removed.

File: python/py-isoduration/Portfile
Comment (reneeotten) [reply python/py-isoduration/Portfile:14]: upstream lists a license both on the PyPI package as well as GitHub: https://github.com/bolsote/isoduration/blob/master/LICENSE

File: python/py-jsonpointer/Portfile
Comment (reneeotten) [inline python/py-jsonpointer/Portfile:15]: please remove

File: python/py-isoduration/Portfile
Comment (johankytt) [reply python/py-isoduration/Portfile:14]: I mis-worded my response. The license seems to be a "custom" license, not an "official" one with a proper name. So the name of the license is missing. Copy-pasting the whole license does not seem reasonable.

Comment (reneeotten) [reply python/py-isoduration/Portfile:15]: `ISC` is a perfectly fine license, there are many other ports in MacPorts that use the same

Comment (johankytt) [reply python/py-isoduration/Portfile:15]: Does ISC stand for some kind of a freeform license? Should it be always, when other licenses are not applicable, used?

Comment (reneeotten) [reply python/py-isoduration/Portfile:16]: no, there is nothing "freeform" about it (it's specified in [SPDX](https://spdx.org/licenses/ISC.html)) and it shouldn't always be used. It should be used when upstream specifies it as their license.

File: python/py-jsonschema/Portfile
Comment (reneeotten) [reply python/py-jsonschema/Portfile:12]: then the `platforms` field should be removed, as `darwin` (without the curly brackets since there is only one thing specified) is the default value.

File: python/py-jsonschema/Portfile
Comment (reneeotten) [inline python/py-jsonschema/Portfile:80]: setting (or changing) a default variant requires to increase the `revision`

Comment (johankytt) [reply python/py-jsonschema/Portfile:80]: It's listed as an "extra" in the pypi package. No point in installing extra files unless you need them. So having the option of installing less clutter is good.

File: python/py-isoduration/Portfile
Comment (johankytt) [reply python/py-isoduration/Portfile:16]: Right, now I see, pypi lists it indeed as ISC license. The license file in the source package has no mention of the license name.

**Comment by pmetzger** [comment]
@reneeotten I doubt @stromnov is ever going to chime in, so we should make a decision on this on our own.

**Comment by barracuda156** [comment]
Wait, let me test, I just noticed this. I can try on Intel and PowerPC both.

File: python/py-jsonschema/Portfile
Comment (pmetzger) [reply python/py-jsonschema/Portfile:79]: @reneeotten I doubt @stromnov is ever going to comment.

**Comment by barracuda156** [comment]
Well, installing `py*-jsonschema` now works, it installs the earlier version of the port normally:
```
macmini:~ svacchanda$ port -v installed py311-jsonschema
The following ports are currently installed:
  py311-jsonschema @4.17.3_0 requested_variants='' platform='darwin 10' archs='noarch' date='2023-04-22T08:35:41+0800'
  py311-jsonschema @4.17.3_1 (active) requested_variants='' platform='darwin 10' archs='noarch' date='2024-01-26T08:53:05+0800'
```

**Comment by pmetzger** [comment]
So are we now ready to merge?

**Comment by barracuda156** [comment]
@pmetzger As for fixing `py-jsonschema`, this solves the problem and closes related ticket. I will not test new ports at the moment, but since CI pass, no reason not to merge.

File: python/py-jsonschema/Portfile
Comment (stromnov) [reply python/py-jsonschema/Portfile:79]: Firstly, I think we could drop EOL support (pre-3.8) of Python.

Secondly, I think it's not a good idea to force the +format_nongpl variant only to implicitly support the Jupyter ecosystem. This can be done in alternative ways: by explicitly writing dependencies in Jupiter or by creating a meta port (for example, py-jsonschema-jupyter).

File: python/py-jsonpointer/Portfile
Comment (pmetzger) [inline python/py-jsonpointer/Portfile:15]: Why add a 3.7 version? It's out of support.

File: python/py-jsonschema/Portfile
Comment (pmetzger) [reply python/py-jsonschema/Portfile:79]: Okay, @johankytt and @barracuda156, let's drop 3.7 etc., and please otherwise follow @stromnov's requests.

**Comment by barracuda156** [comment]
Could we please merge this? Turns out that `mbedtls3` depends on `py-jsonschema`, and now it tries to pull in ridiculous dependencies due to Rust.

**Comment by barracuda156** [comment]
In combo with a simultaneous switch to python312 it is quite bad now, even `-n` does not work:
```
36-246% sudo port -v -n upgrade mbedtls3
Password:
Warning: configured user/group macports does not exist, will build as root
--->  Computing dependencies for mbedtls3.....
Error: Cannot install py312-rpds-py for the arch 'ppc' because
Error: its dependency py312-maturin only supports the archs 'arm64 x86_64'.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
```

**Comment by pmetzger** [comment]
@barracuda156 @stromnov commented (look through the diffs) and asked for some explicit changes to the pull request.

**Comment by barracuda156** [comment]
@pmetzger Got it. Gonna wait then.

File: python/py-jsonschema/Portfile
Comment (johankytt) [reply python/py-jsonschema/Portfile:79]: How do you create a meta port that installs the jsonschema variant if direct variant dependency is not possible? Or do you mean the complete duplication of the jsonschema+nongpl port? I think I've seen some kind of a meta port somewhere before, but can't find an example at the moment.

**Comment by pmetzger** [comment]
@johankytt Ping.

**Comment by johankytt** [comment]
> @johankytt Ping.

I posted a question a couple of days ago, haven't received an answer yet. Check higher up on the page, in the change request discussion.

**Comment by barracuda156** [comment]
Just in case, if the controversial matter are new ports (well, there should be nothing controversial in pegging a port to a working version to unbreak it, and it is a fairly common practice in such cases), maybe split the PR and merge a fix for jsonschema first? Upstream considers making it optional, but it won’t happen overnight.

**Comment by stromnov** [comment]
Please split the PR into two parts: dependencies and the py-jsonschema itself. It's quite difficult to process this PR at once.

**Comment by stromnov** [comment]
My variant of updated `py-jsonschema` with `py-jsonschema-format-nongpl` metaport - https://github.com/stromnov/macports-ports/commit/3ab220653ced5416007709da9a62dec51f63e11c

**Comment by pmetzger** [comment]
@stromnov Would you like to do the rest of this as well? You are both the maintainer and have commit access after all.

File: python/py-fqdn/Portfile
Comment (stromnov) [inline python/py-fqdn/Portfile:29]: Probably not needed (only for python<3.8).
https://github.com/ypcrts/fqdn/blob/3a4feda3414bde7b3386e995ed1d2ed36607d4cf/setup.py#L17

File: python/py-sympy/Portfile
Comment (stromnov) [inline python/py-sympy/Portfile:1]: AFAIK, already at 1.12 (https://github.com/macports/macports-ports/commit/f30a8eac572b322cb77e843c4086dad23e96fa93)

**Comment by johankytt** [comment]
> My variant of updated `py-jsonschema` with `py-jsonschema-format-nongpl` metaport - [stromnov@3ab2206](https://github.com/stromnov/macports-ports/commit/3ab220653ced5416007709da9a62dec51f63e11c)

So can this pull request be deleted then? It doesn't look like your version fixes the issue for older operating systems.

**Comment by barracuda156** [comment]
@johankytt I guess we still need your PR to finally fix builds on older systems. Maybe split the fix from the rest, as suggested above? We need the fallback, and that should be uncontroversial to merge.

**Comment by johankytt** [comment]
> @johankytt I guess we still need your PR to finally fix builds on older systems. Maybe split the fix from the rest, as suggested above? We need the fallback, and that should be uncontroversial to merge.

Ok, I'll wait until that commit gets merged to master and then try to rebase my update onto that.

**Comment by barracuda156** [comment]
@johankytt @stromnov What is the status on this?
I really need `mbedtls3` building, several ports of mine use it, and everything is broken now due to this pulled Rust dependency.

**Comment by pmetzger** [comment]
@barracuda156 Feel free to do your own pull request with the pieces you need.

**Comment by johankytt** [comment]
It seems that the commit @stromnov made still doesn't appear in the master branch. As it also doesn't appear in my fork I don't see a way how to rebase my commits on top of that. As a heads up, I will offline for a month starting from Thursday evening and quite busy until then, so don't expect major updates from me. For small stuff I can probably find time. If someone wants to take this over, I can easily delete the pull request.

**Comment by pmetzger** [comment]
image
@johankytt It's on master.

**Comment by barracuda156** [comment]
I can make a fallback part as a separate PR, that is perhaps independent of other stuff.

**Comment by johankytt** [comment]
sympy has nothing to do with it. This is the commit I'm waiting for:
https://github.com/stromnov/macports-ports/commit/3ab220653ced5416007709da9a62dec51f63e11c

**Comment by pmetzger** [comment]
@johankytt That's not part of this repository though.

**Comment by johankytt** [comment]
Update based on @stromnov's example commit. I left the old py versions on purpose as I don't have time to check whether something breaks or not if they are removed. This all happened thanks to a delayed flight. Now I am offline for a month, so if there are any more issues, then feel free to hijack or discard this PR.

**Comment by pmetzger** [comment]
@johankytt You never resolved the conflicts.

@barracuda156 If you want to get this done quickly, I suggest you take it up yourself.

**Comment by barracuda156** [comment]
@pmetzger I believe, what I needed to be fixed has been done by @stromnov in https://github.com/macports/macports-ports/commit/ff84bb388a039ee1b94fdd7419cddd020a4fc01c

**Comment by pmetzger** [comment]
@barracuda156 Is there anything else from this pull request that is still important at the moment, or can it be closed?

**Comment by barracuda156** [comment]
@pmetzger Nothing for me, at least at the moment. My only concern was broken `mbedtls3` due to Rust being pulled by `py-jsonschema`. That has been addressed.

**Comment by pmetzger** [comment]
Closing this for now. If there are pieces that need to be resurrected, we can do that later.

**Pull request closed without merging**
